maven-xml-impl/src/main/java/org/apache/maven/internal/xml/XmlNodeWriter.java

```diff
@@ -18,49 +18,146 @@
  */
 package org.apache.maven.internal.xml;
 
-import java.io.PrintWriter;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamWriter;
+
 import java.io.Writer;
 import java.util.Map;
 
 import org.apache.maven.api.xml.XmlNode;
-import org.codehaus.plexus.util.xml.PrettyPrintXMLWriter;
-import org.codehaus.plexus.util.xml.XMLWriter;
+import org.codehaus.stax2.util.StreamWriterDelegate;
 
 /**
  *
  */
 public class XmlNodeWriter {
     public static void write(Writer writer, XmlNode dom) {
-        write(new PrettyPrintXMLWriter(writer), dom);
-    }
-
-    public static void write(PrintWriter writer, XmlNode dom) {
-        write(new PrettyPrintXMLWriter(writer), dom);
-    }
-
-    public static void write(XMLWriter xmlWriter, XmlNode dom) {
-        write(xmlWriter, dom, true);
+        try {
+            XMLOutputFactory factory = new com.ctc.wstx.stax.WstxOutputFactory();
+            factory.setProperty(XMLOutputFactory.IS_REPAIRING_NAMESPACES, false);
+            factory.setProperty(com.ctc.wstx.api.WstxOutputProperties.P_USE_DOUBLE_QUOTES_IN_XML_DECL, true);
+            factory.setProperty(com.ctc.wstx.api.WstxOutputProperties.P_ADD_SPACE_AFTER_EMPTY_ELEM, true);
+            XMLStreamWriter serializer = new IndentingXMLStreamWriter(factory.createXMLStreamWriter(writer));
+            write(serializer, dom);
+            serializer.close();
+        } catch (XMLStreamException e) {
+            throw new RuntimeException("Unexpected error when writing XML", e);
+        }
     }
 
-    public static void write(XMLWriter xmlWriter, XmlNode dom, boolean escape) {
-        // TODO: move to XMLWriter?
-        xmlWriter.startElement(dom.getName());
+    public static void write(XMLStreamWriter xmlWriter, XmlNode dom) throws XMLStreamException {
+        xmlWriter.writeStartElement(dom.getPrefix(), dom.getName(), dom.getNamespaceUri());
         for (Map.Entry<String, String> attr : dom.getAttributes().entrySet()) {
-            xmlWriter.addAttribute(attr.getKey(), attr.getValue());
+            xmlWriter.writeAttribute(attr.getKey(), attr.getValue());
         }
         for (XmlNode aChildren : dom.getChildren()) {
-            write(xmlWriter, aChildren, escape);
+            write(xmlWriter, aChildren);
         }
-
         String value = dom.getValue();
         if (value != null) {
-            if (escape) {
-                xmlWriter.writeText(value);
-            } else {
-                xmlWriter.writeMarkup(value);
+            xmlWriter.writeCharacters(value);
+        }
+        xmlWriter.writeEndElement();
+    }
+
+    static class IndentingXMLStreamWriter extends StreamWriterDelegate {
+
+        int depth = 0;
+        boolean hasChildren = false;
+        boolean anew = true;
```

Comment:
do you really need this? If so, it might need to be set to true when entering writeStartDocument

Reply:
The idea is only to be able to not ring a `\n` at the beginning when printing a snippet and not a full document.  And that's for the `XmlNodeImpl#toString()` call, once again.  Without this flag and check, we end up with a leading `\n` at the beginning: `\n<configuration></configuration>` ...